LibOS/shim/src/ipc/shim_ipc_nsimpl.h

```diff
@@ -265,37 +270,46 @@ static int __add_range (struct range * r, int off, IDTYPE owner,
     ret = __set_range_bitmap(off, false);
     if (ret == -EEXIST) {
         struct range * tmp;
-        struct hlist_node * pos;
 
-        hlist_for_each_entry(tmp, pos, head, hlist)
+        listp_for_each_entry(tmp, head, hlist)
             if (tmp->offset == off) {
-                hlist_del(&tmp->hlist);
-                list_del(&tmp->list);
+                listp_del(tmp, head, hlist);
+
+                /* Chia-Che Tsai 10/17/17: only when tmp->owner is non-NULL,
+                 * and tmp->owner->vmid == cur_process.vmid, tmp is on the
+                 * owned list, otherwise it is an offered. */
+                if (tmp->owner && tmp->owner->vmid == cur_process.vmid)
```

Comment:
@donporter Please check this change.

I believe the right logic is that only when owner is non-NULL and owner->vmid is not the current process, the range belongs to the owned list (owned by the current process). The following code should point out pretty clear.

Reply:
I think this change is good and consistent with other code.  Thanks!